File: ISP/isp.ts
interface Printable {
    print(): void;
}

interface Scannable {
    scan(): void;
}

interface Faxable {
    fax(): void;
}

interface Photocopiable {
    photocopy(): void;
}

interface Emailable {
    email(): void;
}

class OldPrinter implements Printable {
    print(): void {
        console.log("Old printer printing document");
    }
}

class MultiFunctionPrinter
    implements Printable, Scannable, Faxable, Photocopiable, Emailable {

    print(): void {
        console.log("Printing document");
    }

    scan(): void {
        console.log("Scanning document");
    }

    fax(): void {
        console.log("Sending fax");
    }

    photocopy(): void {
        console.log("Photocopying document");
    }

    email(): void {
        console.log("Emailing document");
    }
}

class PrintService {
    static startPrinting(printer: Printable): void {
        printer.print();
    }
}

class ScanService {
    static startScanning(scanner: Scannable): void {
        scanner.scan();
    }
}

class isp {
    static main(): void {
        const oldPrinter: Printable = new OldPrinter();
        const modernPrinter: Printable = new MultiFunctionPrinter();

        console.log("Printing");
        PrintService.startPrinting(oldPrinter);
        PrintService.startPrinting(modernPrinter);

        console.log("\nScanning");
        const scanner: Scannable = new MultiFunctionPrinter();
        ScanService.startScanning(scanner);
    }
}

isp.main();

File: ISP/isp.java
interface Printable {
   void print();
}
interface Scannable {
   void scan();
}
interface Faxable {
   void fax();
}
interface Photocopiable {
   void photocopy();
}


interface Emailable {
   void email();
}


class OldPrinter implements Printable {


   public void print() {
       System.out.println("Old printer printing document");
   }
}


class MultiFunctionPrinter implements Printable, Scannable, Faxable, Photocopiable, Emailable {


   public void print() {
       System.out.println("Printing document");
   }


   public void scan() {
       System.out.println("Scanning document");
   }


   public void fax() {
       System.out.println("Sending fax");
   }


   public void photocopy() {
       System.out.println("Photocopying document");
   }


   public void email() {
       System.out.println("Emailing document");
   }
}


class PrintService {
   public static void startPrinting(Printable printer) {
       printer.print();
   }
}


class ScanService {
   public static void startScanning(Scannable scanner) {
       scanner.scan();
   }
}


public class isp {
   public static void main(String[] args) {
       Printable oldPrinter = new OldPrinter();
       Printable modernPrinter = new MultiFunctionPrinter();
       System.out.println("Printing");
       PrintService.startPrinting(oldPrinter);
       PrintService.startPrinting(modernPrinter);
       System.out.println("\nScanning");
       Scannable scanner = new MultiFunctionPrinter();
       ScanService.startScanning(scanner);
   }
}
